process/resultsParser/resultsParser.go

```diff
@@ -0,0 +1,401 @@
+		t := findFirstOrNoneTopic(e.Topics, func(topic []byte) []byte {
+			if string(topic) == base64Address {
+				return topic
+			}
+			return nil
+		})
+
+		if t != nil {
+			return e
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to find single or none event: %w", err)
+	}
+
+	if event == nil {
+		return nil, nil
+	}
+
+	returnCode, returnDataParts, err := sliceDataFieldInParts(string(event.Data))
+	if err != nil {
+		return nil, fmt.Errorf("failed to slice data field in parts: %w", err)
+	}
+
+	return &ResultOutcome{
+		ReturnCode:    *returnCode,
+		ReturnMessage: returnCode.String(),
+		Values:        returnDataParts,
+	}, nil
+}
+
+func parseOutcomeWithFallbackHeuristics(tx *transaction.ApiTransactionResult, receiver string) (*ResultOutcome, error) {
+	for _, resultItem := range tx.SmartContractResults {
+		event, findErr := findSingleOrNoneEvent(resultItem.Logs, OnWriteLog, func(e *transaction.Events) *transaction.Events {
+			addressIsSender := e.Address == tx.Sender
+			firstTopicIsContract := false
+			if e.Topics[0] != nil {
+				decodeString, topicDecode := hex.DecodeString(string(e.Topics[0]))
+				if topicDecode != nil {
+					return nil
+				}
+				firstTopicIsContract = string(decodeString) == receiver
+
+				if firstTopicIsContract && addressIsSender {
+					return e
+				}
+			}
+
+			return nil
+		})
+		if findErr != nil {
+			return nil, findErr
+		}
+
+		if event == nil {
+			return nil, nil
+		}
+
+		returnCode, returnDataParts, sliceErr := sliceDataFieldInParts(string(event.Data))
+		if sliceErr != nil {
+			return nil, sliceErr
+		}
+
+		return &ResultOutcome{
+			ReturnCode:    *returnCode,
+			ReturnMessage: returnCode.String(),
+			Values:        returnDataParts,
+		}, nil
+	}
+
+	return nil, nil
+}
+
+func sliceDataFieldInParts(data string) (*ReturnCode, []*bytes.Buffer, error) {
+	if data == "" {
+		return nil, nil, ErrEmptyDataField
+	}
+
+	// By default, skip the first part, which is usually empty (e.g. "[empty]@6f6b")
+	startingIndex := 1
+
+	// Before trying to parse the hex strings, cut the unwanted parts of the data field, in case of token transfers:
+	if strings.HasPrefix(data, "ESDTTransfer") {
+		// Skip "ESDTTransfer" (1), token identifier (2), amount (3)
+		startingIndex = 3
+	} else {
+		// TODO: Upon gathering more transaction samples, fix for other kinds of transfers, as well (future PR, as needed).
+	}
+
+	// TODO: make this a function that returns a slice of bytes
+	parts := stringToBuffers(data)
+	if len(parts) <= startingIndex {
+		return nil, nil, ErrCannotProcessDataField
+	}
+
+	returnCodePart := parts[startingIndex]
+	returnDataParts := parts[startingIndex+1:]
+
+	if returnCodePart.Len() == 0 {
+		return nil, nil, ErrNoReturnCode
+	}
+
+	returnCode := fromBuffer(*returnCodePart)
+	return &returnCode, returnDataParts, nil
+}
+
+func stringToBuffers(joinedString string) []*bytes.Buffer {
```

Comment:
why do you need a slice of buffer, instead of [][]byte ?

Reply:
changed.